agents/agents/guard/fraud.go

```diff
@@ -0,0 +1,504 @@
+package guard
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	ethTypes "github.com/ethereum/go-ethereum/core/types"
+	"github.com/synapsecns/sanguine/agents/types"
+)
+
+// handleSnapshot checks a snapshot for invalid states.
+// If an invalid state is found, initiate slashing and submit a state report.
+func (g Guard) handleSnapshot(ctx context.Context, log ethTypes.Log) error {
+	fraudSnapshot, err := g.inboxParser.ParseSnapshotAccepted(log)
+	if err != nil {
+		return fmt.Errorf("could not parse snapshot accepted: %w", err)
+	}
+
+	// Verify each state in the snapshot.
+	for stateIndex, state := range fraudSnapshot.Snapshot.States() {
+		isSlashable, err := g.isStateSlashable(ctx, state, fraudSnapshot.Agent)
+		if err != nil {
+			return fmt.Errorf("could not handle state: %w", err)
+		}
+		if !isSlashable {
+			continue
+		}
+
+		// Initiate slashing on origin.
+		_, err = g.domains[state.Origin()].LightInbox().VerifyStateWithSnapshot(
+			ctx,
+			g.unbondedSigner,
+			int64(stateIndex),
+			fraudSnapshot.Payload,
+			fraudSnapshot.Signature,
+		)
+		if err != nil {
+			return fmt.Errorf("could not verify state with snapshot: %w", err)
+		}
+
+		// Check if we should submit the state report.
+		shouldSubmit, err := g.shouldSubmitStateReport(ctx, fraudSnapshot)
+		if err != nil {
+			return fmt.Errorf("could not check if should submit state report: %w", err)
+		}
+		if !shouldSubmit {
+			return nil
+		}
+
+		// Submit the state report.
+		srSignature, _, _, err := state.SignState(ctx, g.bondedSigner)
+		if err != nil {
+			return fmt.Errorf("could not sign state: %w", err)
+		}
+		_, err = g.domains[g.summitDomainID].Inbox().SubmitStateReportWithSnapshot(
```

Comment:
fyi, this will deref if g.summitDomainID set incorrectly

Reply:
Now checking for this in constructor, so we know we will have it.